src/components/TodoProvider.tsx

```diff
@@ -0,0 +1,116 @@
+import React, { createContext, useEffect, useState } from 'react';
+
+import { Todo } from '../types/Todo';
+import {
+  addTodo,
+  deleteTodo,
+  getTodos,
+  updateTodo,
+} from '../api/todos';
+import { USER_ID } from '../utils/constants';
+
+interface TodoContextProps {
+  todos: Todo[];
+  addTodoHandler: (todo: Omit<Todo, 'id'>) => Promise<void>;
+  deleteTodoHandler: (todoId: number) => Promise<void>;
+  updateTodoHandler: (todo: Todo, property: Partial<Todo>) => Promise<void>;
+  errorMessage: string;
+  setErrorMessage: (str: string) => void;
+  isLoadingMap: {};
+}
+
+export const TodoContext = createContext<TodoContextProps>({
+  todos: [],
+  addTodoHandler: async () => { },
+  deleteTodoHandler: async () => { },
+  updateTodoHandler: async () => { },
+  errorMessage: '',
+  setErrorMessage: () => { },
+  isLoadingMap: {},
+});
+
+type TodoProviderProps = {
+  children: React.ReactNode;
+};
+
+export const TodoProvider: React.FC<TodoProviderProps> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [isLoadingMap, setIsLoadingMap]
+    = useState<{ [key: number]: boolean } | {}>({});
```

Comment:
You can just store an array of loading todos `ids` and it will be a bit easy to manage